Meme/Meme/NetworkClient.swift

```diff
@@ -0,0 +1,86 @@
+//
+//  NetworkClient.swift
+//  Meme
+//
+//  Created by Mesfin Bekele Mekonnen on 7/29/19.
+//  Copyright © 2019 Mesfin Bekele Mekonnen. All rights reserved.
+//
+
+import Foundation
+
+class NetworkClient {
+
+    let templateParser: TemplateParser
+    let memeParser: MemeParser
+    let templateExampleParser: TemplateExampleParser
+    var templates = [Template]()
+    var templateExample = TemplateExample(name: "", exampleURL: URL(string: "https://www.apple.com")!)
+    var meme = Meme(name: "", imageURL: URL(string: "https://www.apple.com")!)
+
+    init(templateParser: TemplateParser, memeParser: MemeParser, templateExampleParser: TemplateExampleParser) {
+        self.templateParser = templateParser
+        self.memeParser = memeParser
+        self.templateExampleParser = templateExampleParser
+    }
+
+    fileprivate func getTemplates(_ data: Data?) -> [Template]? {
+        if let data = data {
+            if let templates = self.templateParser.parse(data: data) {
+                return templates
+            }
+        }
+        return nil
+    }
+
+    fileprivate func getTemplateExample(_ data: Data?) -> TemplateExample? {
+        if let data = data {
+            if let templateExample = self.templateExampleParser.parse(data: data) {
+                return templateExample
+            }
+        }
+        return nil
+    }
+
+    fileprivate func getMemes(_ data: Data?, memeName: String) -> Meme? {
+        if let data = data {
+            if let meme = self.memeParser.parse(imageData: data, memeName: memeName) {
+                return meme
+            }
+        }
+        return nil
+    }
+
+    func requestTemplates()  {
+        let session = URLSession(configuration: URLSessionConfiguration.default)
+        let url =  URL(string: "https://memegen.link/api/templates")!
```

Comment:
let's pull this out into an endpoint builder like this maybe?
```swift
enum MemeAPIEndpoint {
    case template
    ...etc
}

struct MemeEndpointBuilder {
    private let baseURLString = "https://memegen.link/api/templates"

    func endpointURL(_ type: MemeAPIEndpoint) -> URL {
        let path: String
        switch type {
        ..do pattern matching and stuff
        }
        return URL(string: baseURLString + path)!
    }
}
```

Reply:
This is nice. I will do this soon.